File: patterns/stack/daily-temperature/daily-temperatures.ts
/**
 * Leetcode 739: Daily Temperatures
 * URL: https://leetcode.com/problems/daily-temperatures
 * Pattern: Stack
 * Stack Type: Monotonic
 * Language: Typescript
 * Insight: Last unresolved element decides next action
 */

function dailyTemperatures(temps: number[]): number[] {
    const n = temps.length;
    const answer = new Array(n).fill(0);
    const stack: number[] = [];

    for (let i = 0; i < n; i++) {
        while(stack.length > 0 && temps[i] > temps[stack[stack.length - 1]]) {
            const prevIndex = stack.pop()!;
            answer[prevIndex] = i - prevIndex
        }
        stack.push(i)
    }

    return answer
}

console.log(dailyTemperatures([73,74,75,71,69,72,76,73]
))

File: patterns/stack/daily-temperature/daily-temperatures.py
# Leetcode 739: Daily Temeprature
# URL: https://leetcode.com/problems/daily-temperatures
# Pattern: Stack
# Stack Type: Monotonic

from typing import List 


def daily_temperatures(temperatures: List[int]) -> List[int]:
    n = len(temperatures)
    answer = [0] * n
    stack = []
    
    for i in range(n):
        while stack and temperatures[i] > temperatures[stack[-1]]:
            prev_index = stack.pop()
            answer[prev_index] = i - prev_index
            
        stack.append(i)
        
    return answer

print(daily_temperatures([73,74,75,71,69,72,76,73]))